mlsdbtool/main.c

```diff
@@ -0,0 +1,180 @@
+/*
+  Copyright (C) 2022 Jolla Ltd.
+  Contact: Daniel Suni <daniel.suni@jolla.com>
+
+  This file is part of geoclue-mlsdb.
+
+  Geoclue-mlsdb is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License.
+*/
+
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#define NEWLINE 10
+#define STDIN 0
+
+/*
+This parser will read MLS full export data in CSV format. Data files can
+be downloaded here: https://location.services.mozilla.com/downloads
+
+The reader assumes the data is numerically sorted by unique cell id and that
+only data lines (i.e. no headers) are passed. To achieve that, run:
+tail -n +2 [CSV file] | sort -t, -k2n,2 -k3n,3 -k4n,4 -k5n,5 -k1,1 | geoclue-mlsdb-tool
+or (preferrably) use the wrapper script.
+
+---
+
+The "network" is allocated as follows:
+0000000000000000000000000000000000000000000000000000000000000000
+        \____ ___/\_______ ______/\_____________ ____________/\/
+             v            v                     v              v
+          NET: 10b Area: 16 bits        Cell ID: 28 bits     Radio: 2b 
+
+The "position" is allocated as 2 concatenated 32-bit floats
+with longitude first and then latitude.
+
+This program will produce 2 files per mcc: .ntw for network data and .loc
+for location data. To finalize the mlsdb data file the two should simply
+be concatenated into a .dat file. (The wrapper will do this too.)
+
+Since the network portion and the data portion are exactly the same size,
+the file is simply split so the the 1st half contains network data, and the
+corresponding location data is found in ${net_data_pos} + ${file_size} / 2
+*/
+uint64_t network;
+uint64_t position;
+uint64_t previous = 0;
+
+// Helper function for debugging (prints out a 64b int as binary)
+void print_bin(uint64_t n) {
+    if (n > 1) {
+        print_bin(n >> 1);
+    }
+    printf("%lu", n & 1);
+}
+
+void add_network(char *str, size_t shift) {
+    uint64_t t = atoi(str);
+    t <<= shift;
+    network |= t;
+}
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
```

Comment:
The cause is that I use a float as if it were an int. This is entirely intentional, because at this point I only care about the ones and zeroes, not the data type.

Reply:
Ah right it was related to that. Well, ok then.